File: referencia/flutter-app/lib/datos/bd/base_datos.dart
/// Base de datos local de CacaoTrace.
///
/// Es LA FUENTE DE VERDAD de la app (RF-SYN-01): todo se escribe aquí primero
/// y la nube es un espejo que se pone al día cuando hay señal. Por eso la app
/// funciona completa en la finca sin cobertura (RNF-01).
library;

import 'package:drift/drift.dart';
import 'package:drift_flutter/drift_flutter.dart';

// El codigo generado vive en un 'part' de este archivo, asi que los enums que
// usan las columnas tienen que estar visibles AQUI, no solo en tablas/.
import '../../nucleo/modelo/etapas.dart';
import 'tablas/apoyo.dart';
import 'tablas/comunes.dart';
import 'tablas/lote.dart';
import 'tablas/produccion.dart';

part 'base_datos.g.dart';

@DriftDatabase(
  tables: [
    // Configuración
    Fincas,
    Equipos,
    UmbralesGuardados,
    Configuracion,
    ModelosIa,
    // Lote de grano
    Lotes,
    Recepciones,
    Aperturas,
    Fermentaciones,
    LecturasFermentacion,
    Volteos,
    Secados,
    LecturasSecado,
    PruebasCorte,
    Sacos,
    InspeccionesAlmacen,
    // Lote de producción
    LotesProduccion,
    LotesProduccionOrigen,
    Tostados,
    Descascarillados,
    Refinados,
    Atemperados,
    Empaques,
    // Apoyo
    MovimientosInventario,
    StocksMinimos,
    Costos,
    Ventas,
    ChecklistsBpm,
    RegistrosBpm,
    Laboratorios,
    Alertas,
    CorreccionesAplicadas,
    Fotos,
    Auditoria,
    ColaSync,
  ],
)
class BaseDatos extends _$BaseDatos {
  BaseDatos([QueryExecutor? ejecutor])
      : super(ejecutor ?? driftDatabase(name: 'cacaotrace'));

  @override
  int get schemaVersion => 1;

  @override
  MigrationStrategy get migration => MigrationStrategy(
        onCreate: (m) async {
          await m.createAll();
          await _crearIndices();
          await _sembrarDatosIniciales();
        },
        beforeOpen: (detalles) async {
          // Las claves foráneas no vienen activadas por defecto en SQLite.
          await customStatement('PRAGMA foreign_keys = ON');
        },
      );

  /// Índices para las consultas que la app hace todo el tiempo.
  ///
  /// Sin estos, la pantalla de inicio recorre tablas enteras en cada refresco,
  /// que con unos pocos lotes no se nota pero con dos años de historial sí.
  Future<void> _crearIndices() async {
    const indices = [
      'CREATE INDEX IF NOT EXISTS idx_lotes_estado ON lotes(estado, eliminado)',
      'CREATE INDEX IF NOT EXISTS idx_lecturas_ferm ON lecturas_fermentacion(fermentacion_id, fecha_hora)',
      'CREATE INDEX IF NOT EXISTS idx_volteos_ferm ON volteos(fermentacion_id, fecha_hora)',
      'CREATE INDEX IF NOT EXISTS idx_lecturas_secado ON lecturas_secado(secado_id, fecha)',
      'CREATE INDEX IF NOT EXISTS idx_alertas_estado ON alertas(estado, fecha)',
      'CREATE INDEX IF NOT EXISTS idx_alertas_dedup ON alertas(clave_dedup)',
      'CREATE INDEX IF NOT EXISTS idx_fotos_lote ON fotos(lote_id, etapa)',
      'CREATE INDEX IF NOT EXISTS idx_cola_proximo ON cola_sync(proximo_intento)',
      'CREATE INDEX IF NOT EXISTS idx_inventario_item ON movimientos_inventario(item, fecha)',
      'CREATE INDEX IF NOT EXISTS idx_costos_lote ON costos(lote_id, lote_produccion_id)',
    ];
    for (final sql in indices) {
      await customStatement(sql);
    }
  }

  /// Datos con los que la app es útil desde el primer arranque.
  Future<void> _sembrarDatosIniciales() async {
    await batch((b) {
      b.insertAll(checklistsBpm, [
        ChecklistsBpmCompanion.insert(
          nombre: 'Limpieza diaria del taller',
          frecuencia: const Value('diario'),
          itemsJson: const Value(
            '["Mesas y superficies limpias y desinfectadas",'
            '"Utensilios lavados y guardados",'
            '"Pisos barridos y trapeados",'
            '"Basura retirada",'
            '"Equipos apagados y limpios"]',
          ),
        ),
        ChecklistsBpmCompanion.insert(
          nombre: 'Higiene personal',
          frecuencia: const Value('diario'),
          itemsJson: const Value(
            '["Manos lavadas antes de empezar",'
            '"Uñas cortas y sin esmalte",'
            '"Cofia o gorro puesto",'
            '"Ropa de trabajo limpia",'
            '"Sin joyas ni reloj",'
            '"Sin heridas expuestas"]',
          ),
        ),
        ChecklistsBpmCompanion.insert(
          nombre: 'Control de plagas',
          frecuencia: const Value('semanal'),
          itemsJson: const Value(
            '["Trampas revisadas",'
            '"Sin rastros de roedores ni insectos",'
            '"Puertas y ventanas con mallas en buen estado",'
            '"Basura fuera del area de proceso"]',
          ),
        ),
        ChecklistsBpmCompanion.insert(
          nombre: 'Agua e insumos',
          frecuencia: const Value('semanal'),
          itemsJson: const Value(
            '["Agua potable disponible",'
            '"Tanque de agua limpio",'
            '"Insumos dentro de su fecha de vencimiento",'
            '"Insumos almacenados sobre pallets, no en el piso"]',
          ),
        ),
      ]);

      b.insertAll(stocksMinimos, [
        StocksMinimosCompanion.insert(item: 'azucar', minimo: 2),
        StocksMinimosCompanion.insert(item: 'manteca', minimo: 0.5),
        StocksMinimosCompanion.insert(item: 'lecitina', minimo: 0.1),
        StocksMinimosCompanion.insert(
            item: 'empaques', minimo: 50, unidad: const Value('unidades')),
      ]);
    });
  }

  /// Borra todo. Solo para tests y para "empezar de cero" desde Ajustes.
  Future<void> vaciar() async {
    await transaction(() async {
      for (final t in allTables) {
        await delete(t).go();
      }
    });
  }
}

File: android/app/src/main/kotlin/ec/cacaotrace/datos/bd/BaseDatos.kt
package ec.cacaotrace.datos.bd

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import androidx.sqlite.db.SupportSQLiteDatabase
import ec.cacaotrace.datos.bd.daos.AlertaDao
import ec.cacaotrace.datos.bd.daos.AperturaDao
import ec.cacaotrace.datos.bd.daos.BpmDao
import ec.cacaotrace.datos.bd.daos.ColaSyncDao
import ec.cacaotrace.datos.bd.daos.ConfiguracionDao
import ec.cacaotrace.datos.bd.daos.CostoDao
import ec.cacaotrace.datos.bd.daos.EquipoDao
import ec.cacaotrace.datos.bd.daos.FermentacionDao
import ec.cacaotrace.datos.bd.daos.FincaDao
import ec.cacaotrace.datos.bd.daos.FotoDao
import ec.cacaotrace.datos.bd.daos.InspeccionAlmacenDao
import ec.cacaotrace.datos.bd.daos.InventarioDao
import ec.cacaotrace.datos.bd.daos.LaboratorioDao
import ec.cacaotrace.datos.bd.daos.LoteDao
import ec.cacaotrace.datos.bd.daos.ProduccionDao
import ec.cacaotrace.datos.bd.daos.PruebaCorteDao
import ec.cacaotrace.datos.bd.daos.RecepcionDao
import ec.cacaotrace.datos.bd.daos.SacoDao
import ec.cacaotrace.datos.bd.daos.SecadoDao
import ec.cacaotrace.datos.bd.daos.VentaDao
import ec.cacaotrace.datos.bd.entidades.AjusteEntidad
import ec.cacaotrace.datos.bd.entidades.AlertaEntidad
import ec.cacaotrace.datos.bd.entidades.AperturaEntidad
import ec.cacaotrace.datos.bd.entidades.AtemperadoEntidad
import ec.cacaotrace.datos.bd.entidades.AuditoriaEntidad
import ec.cacaotrace.datos.bd.entidades.ChecklistBpmEntidad
import ec.cacaotrace.datos.bd.entidades.CorreccionAplicadaEntidad
import ec.cacaotrace.datos.bd.entidades.CostoEntidad
import ec.cacaotrace.datos.bd.entidades.DescascarilladoEntidad
import ec.cacaotrace.datos.bd.entidades.EmpaqueEntidad
import ec.cacaotrace.datos.bd.entidades.EquipoEntidad
import ec.cacaotrace.datos.bd.entidades.FermentacionEntidad
import ec.cacaotrace.datos.bd.entidades.FincaEntidad
import ec.cacaotrace.datos.bd.entidades.FotoEntidad
import ec.cacaotrace.datos.bd.entidades.InspeccionAlmacenEntidad
import ec.cacaotrace.datos.bd.entidades.LaboratorioEntidad
import ec.cacaotrace.datos.bd.entidades.LecturaFermentacionEntidad
import ec.cacaotrace.datos.bd.entidades.LecturaSecadoEntidad
import ec.cacaotrace.datos.bd.entidades.LoteEntidad
import ec.cacaotrace.datos.bd.entidades.LoteProduccionEntidad
import ec.cacaotrace.datos.bd.entidades.LoteProduccionOrigenEntidad
import ec.cacaotrace.datos.bd.entidades.ModeloIaEntidad
import ec.cacaotrace.datos.bd.entidades.MovimientoInventarioEntidad
import ec.cacaotrace.datos.bd.entidades.OperacionSyncEntidad
import ec.cacaotrace.datos.bd.entidades.PruebaCorteEntidad
import ec.cacaotrace.datos.bd.entidades.RecepcionEntidad
import ec.cacaotrace.datos.bd.entidades.RefinadoEntidad
import ec.cacaotrace.datos.bd.entidades.RegistroBpmEntidad
import ec.cacaotrace.datos.bd.entidades.SacoEntidad
import ec.cacaotrace.datos.bd.entidades.SecadoEntidad
import ec.cacaotrace.datos.bd.entidades.StockMinimoEntidad
import ec.cacaotrace.datos.bd.entidades.TostadoEntidad
import ec.cacaotrace.datos.bd.entidades.UmbralGuardadoEntidad
import ec.cacaotrace.datos.bd.entidades.VentaEntidad
import ec.cacaotrace.datos.bd.entidades.VolteoEntidad
import java.util.UUID

/**
 * Base de datos local de CacaoTrace.
 *
 * Es LA FUENTE DE VERDAD de la app (RF-SYN-01): todo se escribe aquí primero y
 * la nube es un espejo que se pone al día cuando hay señal. Por eso la app
 * funciona completa en la finca sin cobertura (RNF-01).
 */
@Database(
    entities = [
        // Configuración
        FincaEntidad::class,
        EquipoEntidad::class,
        UmbralGuardadoEntidad::class,
        AjusteEntidad::class,
        ModeloIaEntidad::class,
        // Lote de grano
        LoteEntidad::class,
        RecepcionEntidad::class,
        AperturaEntidad::class,
        FermentacionEntidad::class,
        LecturaFermentacionEntidad::class,
        VolteoEntidad::class,
        SecadoEntidad::class,
        LecturaSecadoEntidad::class,
        PruebaCorteEntidad::class,
        SacoEntidad::class,
        InspeccionAlmacenEntidad::class,
        // Lote de producción
        LoteProduccionEntidad::class,
        LoteProduccionOrigenEntidad::class,
        TostadoEntidad::class,
        DescascarilladoEntidad::class,
        RefinadoEntidad::class,
        AtemperadoEntidad::class,
        EmpaqueEntidad::class,
        // Apoyo
        MovimientoInventarioEntidad::class,
        StockMinimoEntidad::class,
        CostoEntidad::class,
        VentaEntidad::class,
        ChecklistBpmEntidad::class,
        RegistroBpmEntidad::class,
        LaboratorioEntidad::class,
        AlertaEntidad::class,
        CorreccionAplicadaEntidad::class,
        FotoEntidad::class,
        AuditoriaEntidad::class,
        OperacionSyncEntidad::class,
    ],
    version = 1,
    exportSchema = true,
)
@TypeConverters(Convertidores::class)
abstract class BaseDatos : RoomDatabase() {

    abstract fun lotes(): LoteDao
    abstract fun fincas(): FincaDao
    abstract fun equipos(): EquipoDao
    abstract fun recepciones(): RecepcionDao
    abstract fun aperturas(): AperturaDao
    abstract fun fermentaciones(): FermentacionDao
    abstract fun secados(): SecadoDao
    abstract fun pruebasCorte(): PruebaCorteDao
    abstract fun sacos(): SacoDao
    abstract fun inspecciones(): InspeccionAlmacenDao
    abstract fun produccion(): ProduccionDao
    abstract fun alertas(): AlertaDao
    abstract fun inventario(): InventarioDao
    abstract fun costos(): CostoDao
    abstract fun ventas(): VentaDao
    abstract fun bpm(): BpmDao
    abstract fun laboratorio(): LaboratorioDao
    abstract fun fotos(): FotoDao
    abstract fun configuracion(): ConfiguracionDao
    abstract fun colaSync(): ColaSyncDao

    companion object {
        private const val NOMBRE = "cacaotrace.db"

        @Volatile
        private var instancia: BaseDatos? = null

        fun obtener(contexto: Context): BaseDatos = instancia ?: synchronized(this) {
            instancia ?: construir(contexto.applicationContext).also { instancia = it }
        }

        private fun construir(contexto: Context): BaseDatos =
            Room.databaseBuilder(contexto, BaseDatos::class.java, NOMBRE)
                .addCallback(SembradoInicial)
                // Sin fallbackToDestructiveMigration a propósito: prefiero que
                // una migración olvidada reviente en desarrollo a que borre en
                // silencio dos años de registros del usuario.
                .build()
    }
}

/**
 * Datos con los que la app es útil desde el primer arranque.
 *
 * Se insertan con SQL directo porque en `onCreate` la base todavía no expone
 * sus DAOs: Room aún está construyéndose.
 */
private object SembradoInicial : RoomDatabase.Callback() {

    override fun onCreate(db: SupportSQLiteDatabase) {
        super.onCreate(db)
        sembrarChecklists(db)
        sembrarStocksMinimos(db)
    }

    private fun comunes(): String {
        val ahora = System.currentTimeMillis()
        return "'${UUID.randomUUID()}', $ahora, $ahora, 'local', '', 'PENDIENTE', 0"
    }

    private fun sembrarChecklists(db: SupportSQLiteDatabase) {
        val checklists = listOf(
            "Limpieza diaria del taller" to Triple(
                "diario",
                listOf(
                    "Mesas y superficies limpias y desinfectadas",
                    "Utensilios lavados y guardados",
                    "Pisos barridos y trapeados",
                    "Basura retirada",
                    "Equipos apagados y limpios",
                ),
                true,
            ),
            "Higiene personal" to Triple(
                "diario",
                listOf(
                    "Manos lavadas antes de empezar",
                    "Uñas cortas y sin esmalte",
                    "Cofia o gorro puesto",
                    "Ropa de trabajo limpia",
                    "Sin joyas ni reloj",
                    "Sin heridas expuestas",
                ),
                true,
            ),
            "Control de plagas" to Triple(
                "semanal",
                listOf(
                    "Trampas revisadas",
                    "Sin rastros de roedores ni insectos",
                    "Puertas y ventanas con mallas en buen estado",
                    "Basura fuera del área de proceso",
                ),
                true,
            ),
            "Agua e insumos" to Triple(
                "semanal",
                listOf(
                    "Agua potable disponible",
                    "Tanque de agua limpio",
                    "Insumos dentro de su fecha de vencimiento",
                    "Insumos almacenados sobre pallets, no en el piso",
                ),
                true,
            ),
        )

        checklists.forEach { (nombre, datos) ->
            val (frecuencia, items, activo) = datos
            val itemsJson = items.joinToString(",", "[", "]") { "\"$it\"" }
            db.execSQL(
                """
                INSERT INTO checklists_bpm
                (id, creadoEn, modificadoEn, usuarioId, dispositivoId, estadoSync,
                 eliminado, nombre, frecuencia, itemsJson, activo)
                VALUES (${comunes()}, ?, ?, ?, ?)
                """.trimIndent(),
                arrayOf(nombre, frecuencia, itemsJson, if (activo) 1 else 0),
            )
        }
    }

    private fun sembrarStocksMinimos(db: SupportSQLiteDatabase) {
        val stocks = listOf(
            Triple("azucar", 2.0, "kg"),
            Triple("manteca", 0.5, "kg"),
            Triple("lecitina", 0.1, "kg"),
            Triple("empaques", 50.0, "unidades"),
        )
        stocks.forEach { (item, minimo, unidad) ->
            db.execSQL(
                """
                INSERT INTO stocks_minimos
                (id, creadoEn, modificadoEn, usuarioId, dispositivoId, estadoSync,
                 eliminado, item, minimo, unidad)
                VALUES (${comunes()}, ?, ?, ?)
                """.trimIndent(),
                arrayOf(item, minimo, unidad),
            )
        }
    }
}
